server/conn.go

```diff
@@ -394,7 +394,19 @@ func (cc *clientConn) writeInitialHandshake(ctx context.Context) error {
 }
 
 func (cc *clientConn) readPacket() ([]byte, error) {
-	return cc.pkt.readPacket()
+	// Default MaxAllowedPacket is 16MB.
+	maxAllowedPacket := uint64(16777216)
+	var err error
+	if cc.ctx != nil {
+		if valStr, _ := cc.ctx.GetSessionVars().GetSystemVar(variable.MaxAllowedPacket); len(valStr) > 0 {
```

Comment:
I believe that there will be a tiny performance regression if we have a short-read only workload like `point-get`. But it's not easy to avoid parsing `valStr` here.

We can think treat it as an enhancement in the future.

Reply:
FWIW, the current way other variables avoid parsing the `valStr` is by using a Getter/Setter like this:

https://github.com/pingcap/tidb/blob/8dc1f6b0926e0bf26f38109fceec7992c878cdde/sessionctx/variable/sysvar.go#L212-L221

We mostly do this for instance scoped variables, but we can technically cache a global like this too.